File: java/1602_FindNearestRightNodeInBinaryTree.java
class Solution {
    public TreeNode findNearestRightNode(TreeNode root, TreeNode u) {
        if (root == null) return null;
        Queue<TreeNode> q = new LinkedList<TreeNode>(){{ add(root); }};
        Queue<TreeNode> nextNodes = new LinkedList<TreeNode>(){{ add(root); }};
        while (!q.isEmpty()) {
            int sz = q.size();
            
            for (int i = 0; i < sz; ++i) {
                TreeNode node = q.poll();
                if (node == u) return q.isEmpty() ? null : q.peek();
                if (node.left != null) nextNodes.add(node.left);
                if (node.right != null) nextNodes.add(node.right);
            }
            
            while (!nextNodes.isEmpty()) q.add(nextNodes.poll());
        }
        return null;
    }
}

File: swift/1602_FindNearestRightNodeInBinaryTree.swift
class Solution {
    func findNearestRightNode(_ root: TreeNode?, _ u: TreeNode?) -> TreeNode? {
        var q = [root]
        while !q.isEmpty {
            var nextNodes = [TreeNode?]()
            var sz = q.count
            for i in 0..<sz {
                var node = q[i]
                if node === u {
                    return i < q.count - 1 ? q[i + 1] : nil
                }
                if node?.left != nil {
                    nextNodes.append(node?.left)
                }
                if node?.right != nil {
                    nextNodes.append(node?.right)
                }
            }
            q = nextNodes
        }
        return nil
        
    }
}
